tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -3853,13 +3863,23 @@ SpirvInstruction *SpirvEmitter::processRWByteAddressBufferAtomicMethods(
     }
   } else {
     auto *value = doExpr(expr->getArg(1));
+
+    spv::Op atomicOp = translateAtomicHlslOpcodeToSpirvOpcode(opcode);
+    if (atomicOp == spv::Op::OpAtomicUMax && baseType->isSignedIntegerType())
+      atomicOp = spv::Op::OpAtomicSMax;
+    if (atomicOp == spv::Op::OpAtomicSMax && baseType->isUnsignedIntegerType())
+      atomicOp = spv::Op::OpAtomicUMax;
+    if (atomicOp == spv::Op::OpAtomicUMin && baseType->isSignedIntegerType())
+      atomicOp = spv::Op::OpAtomicSMin;
+    if (atomicOp == spv::Op::OpAtomicSMin && baseType->isUnsignedIntegerType())
+      atomicOp = spv::Op::OpAtomicUMin;
+
     SpirvInstruction *originalVal = spvBuilder.createAtomicOp(
-        translateAtomicHlslOpcodeToSpirvOpcode(opcode),
-        astContext.UnsignedIntTy, ptr, spv::Scope::Device,
+        atomicOp, baseType, ptr, spv::Scope::Device,
```

Comment:
The problem is that the type for `baseType` does not match the type for `value`. That causes a validation error. You cannot cast the pointer to another type, but you could cast the input type to baseType, when they differ, and are the same number of bits. Do not change the opcode. The opcode seem correct as is.

Reply:
Take this example: https://godbolt.org/z/3sz6MEx9s.

It produces:

```
         %39 = OpAccessChain %_ptr_StorageBuffer_uint %buffer %uint_0 %37
         %40 = OpLoad %int %inputValue
         %41 = OpAtomicSMax %uint %39 %uint_1 %uint_0 %40
```

This is invalid because the pointer `%39` points to a uint, and the value `%40` is `%int`. However you can make it valid by changing it to:

```
         %39 = OpAccessChain %_ptr_StorageBuffer_uint %buffer %uint_0 %37
         %40 = OpLoad %int %inputValue
         %bc = OpBitcast %uint %40
         %41 = OpAtomicSMax %uint %39 %uint_1 %uint_0 %bc
```